File: data structures and algorithms/java/06_117.java
import java.util.*;

class HuffmanNode implements Comparable<HuffmanNode> {
    int frequency;
    char character;
    HuffmanNode left;
    HuffmanNode right;
    
    public int compareTo(HuffmanNode node) {
        return frequency - node.frequency;
    }
}

import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

class HuffmanNode implements Comparable<HuffmanNode> {
   char character;
   int frequency;
   HuffmanNode left;
   HuffmanNode right;
   
   @Override
   public int compareTo(HuffmanNode node) {
      return this.frequency - node.frequency;
   }
}

public class HuffmanCoding {
   public static void main(String[] args) {
      String text = "this is an example for huffman encoding";
      
      // Count character frequencies
      Map<Character, Integer> frequency = new HashMap<>();
      for (char c : text.toCharArray()) {
         frequency.put(c, frequency.getOrDefault(c, 0) + 1);
      }
      
      // Create priority queue of nodes
      PriorityQueue<HuffmanNode> queue = 
            new PriorityQueue<>();
      for (Map.Entry<Character, Integer> entry : 
            frequency.entrySet()) {
         HuffmanNode node = new HuffmanNode();
         node.character = entry.getKey();
         node.frequency = entry.getValue();
         node.left = null;
         node.right = null;
         queue.add(node);
      }
      
      // Build Huffman Tree
      HuffmanNode root = null;
      while (queue.size() > 1) {
         // Extract two nodes with lowest frequency
         HuffmanNode x = queue.poll();
         HuffmanNode y = queue.poll();
         
         // Create new internal node
         HuffmanNode sum = new HuffmanNode();
         sum.frequency = x.frequency + y.frequency;
         sum.character = '-';
         sum.left = x;
         sum.right = y;
         root = sum;
         queue.add(sum);
      }
      
      // Generate codes for each character
      Map<Character, String> codes = new HashMap<>();
      printCode(root, "", codes);
      
      // Display generated codes
      System.out.println("Huffman Codes:");
      for (Map.Entry<Character, String> entry : 
            codes.entrySet()) {
         System.out.println(entry.getKey() + ": " + 
               entry.getValue());
      }
      
      // Encode the original text
      StringBuilder encoded = new StringBuilder();
      for (char c : text.toCharArray()) {
         encoded.append(codes.get(c));
      }
      
      System.out.println("\nEncoded text: " + 
            encoded.toString());
      
      // Calculate compression ratio
      double originalSize = text.length() * 8; // 8 bits per char
      double compressedSize = encoded.length();
      System.out.printf("Compression ratio: %.2f\n", 
            compressedSize / originalSize);
   }
   
   /**
    * Recursively generate Huffman codes
    */
   private static void printCode(HuffmanNode root, String s, 
         Map<Character, String> codes) {
      // Leaf node (actual character)
      if (root.left == null && root.right == null && 
            Character.isLetter(root.character)) {
         codes.put(root.character, s);
         return;
      }
      
      // Traverse left (add 0) and right (add 1)
      printCode(root.left, s + "0", codes);
      printCode(root.right, s + "1", codes);
   }
}

File: data structures and algorithms/cs/06_117.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class HuffmanNode : IComparable<HuffmanNode>
{
   public char Character { get; set; }
   public int Frequency { get; set; }
   public HuffmanNode Left { get; set; }
   public HuffmanNode Right { get; set; }

   public int CompareTo(HuffmanNode other)
   {
      return this.Frequency.CompareTo(other.Frequency);
   }
}

public class MinHeap<T> where T : IComparable<T>
{
   private List<T> items = new List<T>();

   public void Enqueue(T item)
   {
      items.Add(item);
      int ci = items.Count - 1;
      while (ci > 0)
      {
         int pi = (ci - 1) / 2;
         if (items[ci].CompareTo(items[pi]) >= 0)
            break;
         T tmp = items[ci];
         items[ci] = items[pi];
         items[pi] = tmp;
         ci = pi;
      }
   }

   public T Dequeue()
   {
      if (items.Count == 0)
         throw new InvalidOperationException();

      T frontItem = items[0];
      items[0] = items[items.Count - 1];
      items.RemoveAt(items.Count - 1);

      if (items.Count > 0)
      {
         int pi = 0;
         while (true)
         {
            int ci = pi * 2 + 1;
            if (ci >= items.Count)
               break;

            if (ci + 1 < items.Count &&
               items[ci + 1].CompareTo(items[ci]) < 0)
               ci++;

            if (items[pi].CompareTo(items[ci]) <= 0)
               break;

            T tmp = items[pi];
            items[pi] = items[ci];
            items[ci] = tmp;
            pi = ci;
         }
      }
      return frontItem;
   }

   public int Count => items.Count;

   public T Peek() =>
      items.Count > 0
         ? items[0]
         : throw new InvalidOperationException();
}

public class HuffmanCoding
{
   public static Dictionary<char, string> BuildHuffmanTree(
      string text)
   {
      // Count frequencies
      var frequency = new Dictionary<char, int>();
      foreach (char c in text)
      {
         if (!frequency.ContainsKey(c))
            frequency[c] = 0;
         frequency[c]++;
      }

      // Create priority queue
      var queue = new MinHeap<HuffmanNode>();
      foreach (var pair in frequency)
      {
         queue.Enqueue(new HuffmanNode
         {
            Character = pair.Key,
            Frequency = pair.Value
         });
      }

      // Build tree
      while (queue.Count > 1)
      {
         var x = queue.Dequeue();
         var y = queue.Dequeue();

         var sum = new HuffmanNode
         {
            Frequency = x.Frequency + y.Frequency,
            Left = x,
            Right = y
         };

         queue.Enqueue(sum);
      }

      // Generate codes
      var codes = new Dictionary<char, string>();
      GenerateCodes(queue.Peek(), "", codes);
      return codes;
   }

   private static void GenerateCodes(
      HuffmanNode node,
      string code,
      Dictionary<char, string> codes)
   {
      if (node == null)
         return;

      if (node.Left == null && node.Right == null)
      {
         codes[node.Character] = code;
         return;
      }

      GenerateCodes(node.Left, code + "0", codes);
      GenerateCodes(node.Right, code + "1", codes);
   }
}
